### PR TITLE
Ldapi unix socket support

### PR DESCRIPTION
**Pull request by zenlord**
This branch adds ldapi:/// connection support to the user_ldap app (both backend code and frontend code)

### CODE REVIEW

**Comment by zenlord** [comment]
Resolved #24497

**Comment by zenlord** [comment]
@blizzz Sorry for my noob question, but is there anything I should or could do to proceed this PR? The codesniffer check fails, but not on a line that I have manipulated. If from now on, it's all out of my hands, that's fine with me ;)

**Comment by zenlord** [comment]
Fixes #24497

**Comment by zenlord** [comment]
@blizzz ping?

**Comment by zenlord** [comment]
@blizzz just wanted to confirm that these patches still work on version 21.0.1.

**Comment by zenlord** [comment]
@blizzz Patches still work with newest version (22.2.3)

**Comment by zenlord** [comment]
@blizzz Patches still apply cleanly with version 22.2.5. System is running smoothly since the PR was created.

@jospoortvliet  Last ping from my end. First time collaborator to an open source project, and very disappointing to see absolutely noone cares.

File: apps/user_ldap/lib/Wizard.php [hunk 1]
```diff
@@ -1349,6 +1363,8 @@ private function getPortSettingsToTry() {
 				$portSettings[] = ['port' => $port, 'tls' => true];
 			}
 			$portSettings[] = ['port' => $port, 'tls' => false];
+		} elseif ($port === 0) {
+			$portSettings[] = ['port' => '', 'tls' => false];
```
Comment (come-nc) [inline apps/user_ldap/lib/Wizard.php:1367]: This should only happen if ldapi is used to avoid surprises

File: apps/user_ldap/lib/Connection.php [hunk 2]
```diff
@@ -424,8 +425,14 @@ private function doCriticalValidation() {
 			(string)$this->configPrefix .'): ';
 
 		//options that shall not be empty
-		$options = ['ldapHost', 'ldapPort', 'ldapUserDisplayName',
+		$options = ['ldapHost', 'ldapUserDisplayName',
 			'ldapGroupDisplayName', 'ldapLoginFilter'];
+
+		//ldapPort should not be empty either unless ldapHost is pointing to a socket
+		if(substr($this->configuration->ldapHost, 0, strlen('ldapi://')) != 'ldapi://') {
```
Comment (come-nc) [inline apps/user_ldap/lib/Connection.php:432]: This test is done in a lot of places, could you add a method on configuration that contains this test?
```suggestion
		if ($this->configuration->usesLdapi()) {
```

**Review by come-nc — COMMENTED** [review]
See comments.
Please also run `composer run cs:fix` for code style.

Does this only work with openldap?

File: apps/user_ldap/lib/Wizard.php [hunk 3]
```diff
@@ -1318,6 +1330,8 @@ private function getConnection() {
 	 * @return array
 	 */
 	private function getDefaultLdapPortSettings() {
+		// LDAPI unix socket support: why are ports 7389 and 389 duplicated here?
+		// LDAPI unix socket support: why are tls values wrong here?
```
Comment (come-nc) [inline apps/user_ldap/lib/Wizard.php:1334]: tls toggle here means starttls. Ports 7636 and 636 are for ldaps which does not use starttls.
Ports 7389 and 389 are first tried with starttls and then without if that fails.

Comment (zenlord) [reply apps/user_ldap/lib/Wizard.php:1334]: OK, removed those comments

File: apps/user_ldap/lib/Wizard.php (hunk 1 above)
Comment (zenlord) [reply apps/user_ldap/lib/Wizard.php:1367]: Agree - will use the same test as used elsewhere in the file: (substr($this->configuration->ldapHost, 0, strlen('ldapi://')) != 'ldapi://')

**Comment by szaimen** [comment]
@zenlord any update here? :)

**Comment by zenlord** [comment]
@szaimen @blizzz @come-nc Finally got around resolving the comments made by @come-nc . Instead of doing a rebase, I have simply created a new clone, applied the original commits as a patch, made the code changes and then force-pushed the result to my fork. I hope this is acceptable, because going through all the merges required from rebasing against Nextcloud 20-something got dull very quickly :/

File: apps/user_ldap/lib/Wizard.php (hunk 1 above)
Comment (zenlord) [reply apps/user_ldap/lib/Wizard.php:1382]: Done: https://github.com/nextcloud/server/pull/24574/commits/b13e69e4ebe87231e073492f177fdb04ed3e746a

File: apps/user_ldap/lib/Connection.php (hunk 2 above)
Comment (zenlord) [reply apps/user_ldap/lib/Connection.php]: Done: https://github.com/nextcloud/server/pull/24574/commits/343ff3f37aed6f9ae147ea3d45c9843e23a84fd9

**Comment by szaimen** [comment]
@zenlord thanks for the update! Can you please make sure to resolve DCO? See https://github.com/nextcloud/server/pull/24574/checks?check_run_id=8159237124 for some steps.

Thanks for your efforts!

**Comment by zenlord** [comment]
@szaimen I think I was successful in doing that, and have pushed another commit to solve an issue raised by one of the checks

**Comment by szaimen** [comment]
> @szaimen I think I was successful in doing that, and have pushed another commit to solve an issue raised by one of the checks

DCO looks good to me now, thanks!

File: apps/user_ldap/lib/Wizard.php [hunk 4]
```diff
@@ -97,7 +98,10 @@ public function __destruct() {
 	 * @throws \Exception
 	 */
 	public function countEntries(string $filter, string $type): int {
-		$reqs = ['ldapHost', 'ldapPort', 'ldapBase'];
+		$reqs = ['ldapHost', 'ldapBase'];
+		if ($this->configuration->usesLdapi() === 0) {
```
Comment (come-nc) [inline apps/user_ldap/lib/Wizard.php:102]:
```suggestion
		if (!$this->configuration->usesLdapi()) {
```

File: apps/user_ldap/lib/Wizard.php [hunk 5]
```diff
@@ -196,11 +200,11 @@ public function countInBaseDN() {
 	 * @return int|bool
 	 */
 	public function countUsersWithAttribute($attr, $existsCheck = false) {
-		if (!$this->checkRequirements(['ldapHost',
-			'ldapPort',
-			'ldapBase',
-			'ldapUserFilter',
-		])) {
+		$reqs = ['ldapHost', 'ldapBase', 'ldapUserFilter'];
+		if ($this->configuration->usesLdapi() === 0) {
```
Comment (come-nc) [inline apps/user_ldap/lib/Wizard.php:204]:
```suggestion
		if (!$this->configuration->usesLdapi()) {
```

**Review by come-nc — CHANGES_REQUESTED** [review]
Please change usesLdapi() return type to bool and adapt if checks accordingly.
Also, remove the config var uses_ldapi unless I missed something and it is actually used?

File: apps/user_ldap/lib/Configuration.php [hunk 6]
```diff
@@ -120,6 +120,7 @@ class Configuration {
 		'ldapDefaultPPolicyDN' => null,
 		'ldapExtStorageHomeAttribute' => null,
 		'ldapMatchingRuleInChainState' => self::LDAP_SERVER_FEATURE_UNKNOWN,
+		'usesLdapi' => null,
```
[the remark this answers is not in the document]
Comment (zenlord) [reply apps/user_ldap/lib/Configuration.php]: Deleted it again - my mistake

**Comment by zenlord** [comment]
@come-nc All comments have been addressed - appreciate the hand-holding :/

File: apps/user_ldap/lib/Configuration.php [hunk 7]
```diff
@@ -463,6 +463,7 @@ public function getDefaults(): array {
 			'ldap_user_avatar_rule' => 'default',
 			'ldap_ext_storage_home_attribute' => '',
 			'ldap_matching_rule_in_chain_state' => self::LDAP_SERVER_FEATURE_UNKNOWN,
+			'uses_ldapi' => 0,
```
Comment (come-nc) [inline apps/user_ldap/lib/Configuration.php:466]:
```suggestion
```

**Review by come-nc — CHANGES_REQUESTED** [review]

**Comment by come-nc** [comment]
@zenlord Apart from above comment, could you squash everything into one commit and fix DCO?
Most likely something like `git rebase -i HEAD~16` and replace all pick but the first one by "s".
And then git commit -s --amend to add the sign-off-by and please DCO bot. Then `git push --force-with-lease`.
You may want to push to a backup branch before just in case.

**Comment by zenlord** [comment]
@come-nc I did perform the commands you shared, but had to manually solve some merge conflicts. I'm not sure whether my latest commit is now the single commit you've asked for?

File: apps/user_ldap/lib/Configuration.php (hunk 7 above)
Comment (zenlord) [reply apps/user_ldap/lib/Configuration.php]: Done

**Comment by zenlord** [comment]
@come-nc Looks like this one is better: single commit and DCO is not complaining.

**Review by come-nc — APPROVED** [review]

**Comment by come-nc** [comment]
25 is feature freezed so this will go into 26, waiting for 25 to branch out before merging into master.

Thank you for this contribution @zenlord !

**Comment by come-nc** [comment]
(actually needs 1 more review so I put back to review label)

**Review by CarlSchwan — APPROVED** [review]

**Review by CarlSchwan — COMMENTED** [review]
Need a rebase (last one hopefully)

**Pull request merged**